### PR TITLE
Updated docs and check_model param

### PR DESCRIPTION
**Pull request by nissu99**
## Summary
This PR adds documentation for the SolaraViz instantiation and adds a new test case to handle arguments and keyword arguments in solaraviz.
## Bug / Issue
 SolaraViz docs #2494 

## Implementation
changes in docs and explicit check in _check_model_params.

## Testing

Screenshot from 2024-11-17 16-30-08

## Additional Notes
<!-- Add any additional information that may be relevant for the reviewers, su

### CODE REVIEW

**Comment by EwoutH** [comment]
Thanks for your PR!

Could you format your PR following our [bug template](https://github.com/projectmesa/mesa/blob/main/.github/PULL_REQUEST_TEMPLATE/bug.md?plain=1)? That makes it readable on its own, helps us review and serves as a  historical reference if we ever need to revisit it.

File: docs/getting_started.md [hunk 1]
```diff
@@ -180,6 +180,16 @@ Mesa now uses a new browser-based visualization system called SolaraViz. This al
...
 
+> **Note:** SolaraViz instantiates new models using `**model_parameters.value`, so all model inputs must be keyword arguments.
+
+Ensure your model's `__init__` method accepts keyword arguments matching the `model_params` keys.
+
+```python
+class MyModel(Model):
+    def __init__(self, n_agents):
+        # Initialize the model with N agents
+```
```
Comment (quaquel) [inline docs/getting_started.md:191]:
```suggestion
    def __init__(self, n_agents=10, seed=None):
        super().__init__(seed=seed)
        # Initialize the model with N agents
```
```

File: docs/migration_guide.md [hunk 2]
```diff
@@ -259,6 +259,17 @@ the import from mesa.experimental. Otherwise here is a list of things you need t
 
 Previously SolaraViz was initialized by providing a `model_cls` and a `model_params`. This has changed to expect a model instance `model`. You can still provide (user-settable) `model_params`, but only if users should be able to change them. It is now also possible to pass in a "reactive model" by first calling `model = solara.reactive(model)`. This is useful for notebook environments. It allows you to pass the model to the SolaraViz Module, but continue to use the model. For example calling `model.value.step()` (notice the extra .value) will automatically update the plots. This currently only automatically works for the step method, you can force visualization updates by calling `model.value.force_update()`.
 
+### Model Initialization with Keyword Arguments
+
+With the introduction of SolaraViz in Mesa 3.0, models are now instantiated using `**model_parameters.value`. This means all inputs for initializing a new model must be keyword arguments. Ensure your model's `__init__` method accepts keyword arguments matching the keys in `model_params`.
+
+```python
+class MyModel(mesa.Model):
+    def __init__(self, n_agents, seed=None):
```
Comment (quaquel) [inline docs/migration_guide.md:268]:
```suggestion
    def __init__(self, n_agents=10, seed=None):
```

File: docs/tutorials/visualization_tutorial.ipynb [hunk 3]
```diff
@@ -118,13 +122,13 @@
...
-   "metadata": {},
    "cell_type": "code",
-   "outputs": [],
    "execution_count": null,
+   "metadata": {},
+   "outputs": [],
    "source": [
     "# Create initial model instance\n",
-    "model1 = MoneyModel(50, 10, 10)\n",
+    "model1 = MoneyModel(n=50, width=10, height=10) #keyword arguments\n",
```
Comment (quaquel) [inline docs/tutorials/visualization_tutorial.ipynb:131]:
```suggestion
    "model = MoneyModel(n=50, width=10, height=10) #keyword arguments\n",
```

File: docs/tutorials/visualization_tutorial.ipynb [hunk 4]
```diff
@@ -197,13 +201,13 @@
...
-   "metadata": {},
    "cell_type": "code",
-   "outputs": [],
    "execution_count": null,
+   "metadata": {},
+   "outputs": [],
    "source": [
     "# Create initial model instance\n",
-    "model1 = MoneyModel(50, 10, 10)\n",
+    "model1 = MoneyModel(n=50, width=10, height=10)\n",
```
Comment (quaquel) [inline docs/tutorials/visualization_tutorial.ipynb:210]:
```suggestion
    "model = MoneyModel(n=50, width=10, height=10)\n",
```

File: docs/tutorials/visualization_tutorial.ipynb [hunk 5]
```diff
@@ -306,56 +312,56 @@
...
```
Comment (quaquel) [inline docs/tutorials/visualization_tutorial.ipynb:321]:
```suggestion
    "model = MoneyModel(n=50, width=10, height=10)\n",
```

File: docs/tutorials/visualization_tutorial.ipynb [hunk 6]
```diff
@@ -306,56 +312,56 @@
...
+      ]
      },
      "metadata": {},
      "output_type": "display_data"
     }
    ],
-   "execution_count": 12
+   "source": [
+    "page = SolaraViz(\n",
+    "    model1,\n",
```
Comment (quaquel) [inline docs/tutorials/visualization_tutorial.ipynb:357]:
```suggestion
    "    model,\n",
```

File: mesa/visualization/solara_viz.py [hunk 7]
```diff
@@ -412,6 +412,10 @@ def _check_model_params(init_func, model_params):
     """
     model_parameters = inspect.signature(init_func).parameters
     for name in model_parameters:
+        if name not in model_parameters:
```
Comment (quaquel) [inline mesa/visualization/solara_viz.py:415]: This check makes no sense. You are already iterating over `model_parameters` so this if will always be `False`

**Review by quaquel — CHANGES_REQUESTED** [review]
I have left a few suggestions and 1 big point. However, some of these suggestions might have knock-on consequences elsewhere in the code. Also, you should most likely have to change the signature of MoneyModel to keyword arguments only.

The big thing is that the check in `solara_viz.py` is not doing what you think it is doing. It would be good to explicitly add a test to `test_solara_viz.py` which checks that the appropriate exception is raised if you try to use SolaraViz with a model that has arguments and keyword arguments.

File: mesa/visualization/solara_viz.py (hunk 7 above)
Comment (nissu99) [reply mesa/visualization/solara_viz.py:415]: oops sorry

Comment (nissu99) [reply mesa/visualization/solara_viz.py:415]: @quaquel do you mean something like this?
Screenshot from 2024-11-19 01-50-43

Comment (quaquel) [reply mesa/visualization/solara_viz.py:415]: I am not intimately familiar with the `inspect` library, but this does make sense. I would advise adding a few tests to test_solara_viz.py to ensure this behaves as intended.

**Comment by nissu99** [comment]
@quaquel ,i have added the test case.

File: mesa/visualization/solara_viz.py [hunk 8]
```diff
@@ -411,6 +411,19 @@ def _check_model_params(init_func, model_params):
         ValueError: If a parameter is not valid for the model's initialization function
     """
     model_parameters = inspect.signature(init_func).parameters
+
+    has_var_positional = any(
```
Comment (tpike3) [inline mesa/visualization/solara_viz.py:415]: @nissu99 If we have to use keyword arguments for solara so it updates when the parameter changes, shouldn't this throw an error is someone uses a positional argument?

**Comment by tpike3** [comment]
@nissu99 Thanks for ding this and the fact you updated tests, migration guide, and tutorial is very cool!

I had one comment and will defer to @quaquel for final review.

File: mesa/visualization/solara_viz.py (hunk 8 above)
Comment (nissu99) [reply mesa/visualization/solara_viz.py:415]: i think right now it takes postional arguments too, we might need to make the changes in the model itself to strictly handle keyword only args, please tell if i am wrong, thank you .

File: docs/tutorials/visualization_tutorial.ipynb [hunk 9]
```diff
@@ -118,13 +122,13 @@
...
-   "metadata": {},
    "cell_type": "code",
-   "outputs": [],
    "execution_count": null,
+   "metadata": {},
+   "outputs": [],
    "source": [
     "# Create initial model instance\n",
-    "model1 = MoneyModel(50, 10, 10)\n",
+    "model = MoneyModel(n=50, width=10, height=10) #keyword arguments\n",
```
Comment (tpike3) [inline docs/tutorials/visualization_tutorial.ipynb:131]: The read the docs build is failing, this needs to be `model1` or you need to change the `Solaraviz`  (6 lines down) call to `model`

**Comment by EwoutH** [comment]
@tpike3 @quaquel what's the state of this PR, and what's steps are needed to get it merged?

**Comment by quaquel** [comment]
I think all that is left is the last point on the failing read the docs built. It should be an easy fix that I had hoped to do today (but life intervened)

**Comment by nissu99** [comment]
@quaquel  @EwoutH  will make the required changes today .

File: docs/tutorials/visualization_tutorial.ipynb (hunk 9 above)
Comment (nissu99) [reply docs/tutorials/visualization_tutorial.ipynb]: made this change .

File: mesa/visualization/solara_viz.py [hunk 10]
```diff
@@ -423,6 +423,19 @@ def _check_model_params(init_func, model_params):
...
+        param.kind == inspect.Parameter.VAR_POSITIONAL
+        for param in model_parameters.values()
+    )
+    has_var_keyword = any(
```
Comment (tpike3) [inline mesa/visualization/solara_viz.py:431]: You can delete this as `VAR_KEYWORD` are supported and what they need to use

File: mesa/visualization/solara_viz.py [hunk 11]
```diff
@@ -423,6 +423,19 @@ def _check_model_params(init_func, model_params):
...
+        param.kind == inspect.Parameter.VAR_KEYWORD
+        for param in model_parameters.values()
+    )
+
+    if has_var_positional and has_var_keyword:
```
Comment (tpike3) [inline mesa/visualization/solara_viz.py:436]: @nissu99 Please change this to 

```
if has_var_positional: 
      raise ValueError("Mesa's visualization requires the use of keyword arguments to ensure the parameters are passed to Solara correctly. Please ensure all model parameters are of form param=value" )
```

File: docs/tutorials/visualization_tutorial.ipynb [hunk 12]
```diff
@@ -366,35 +372,35 @@
...
+       "Cannot show ipywidgets in text"
+      ]
      },
      "metadata": {},
      "output_type": "display_data"
     }
    ],
-   "execution_count": 13
+   "source": [
+    "Histogram(model1)"
```
Comment (tpike3) [inline docs/tutorials/visualization_tutorial.ipynb:402]: needs to be `model`

File: docs/tutorials/visualization_tutorial.ipynb (hunk 3 above)
Comment (tpike3) [reply docs/tutorials/visualization_tutorial.ipynb:131]: Needs to be `model` not `model1`

**Review by tpike3 — CHANGES_REQUESTED** [review]
@nissu99 Thanks again for doing this! As indicated in #2535 this is needed

It is passing the build but I found some other `model1` that need to be changed to `model` 

For the solara_viz I would get rid of `VAR_KEYWORD` all together. I also made the value error more explicit as we have some users new to python and coding. Feel free to further improve

As soon as you get this done we will get it merged!

File: docs/tutorials/visualization_tutorial.ipynb (hunk 12 above)
Comment (nissu99) [reply docs/tutorials/visualization_tutorial.ipynb]: build is failing for model

**Comment by nissu99** [comment]
@tpike3  thanks for all the instructive comments ,i have made the changes.

**Review by tpike3 — APPROVED** [review]
Thanks @nissu99!

**Comment by EwoutH** [comment]
Great work, thanks!

**Comment by EwoutH** [comment]
@tpike3 could you please **Squash and merge** when merging PRs? That keeps our commits on the main branch clean.

image

**Comment by Spartan-71** [comment]
Hey @nissu99 @tpike3 @EwoutH @quaquel I was trying to the [Introductory Tutorial](https://mesa.readthedocs.io/latest/tutorials/intro_tutorial.html), but while running the visualization it is throwing an error.

**Comment by Spartan-71** [comment]
Hey @nissu99 @tpike3 @EwoutH @quaquel I was trying to the [Introductory Tutorial](https://mesa.readthedocs.io/latest/tutorials/intro_tutorial.html), but while running the visualization it is throwing an error.
image
 Even though I have mentioned `n` in the `model_parameter`.
Am I missing something or this is an actual bug ?

**Comment by quaquel** [comment]
1. Which version of mesa do you have?
2. Exactly where in the tutorial do you run into an issue?

**Comment by Spartan-71** [comment]
@quaquel  
I'm using mesa version `3.1.3`  and facing this issue in `Part 1 - Basic Dashboard` under [Visualization Tutorial.](https://mesa.readthedocs.io/latest/tutorials/visualization_tutorial.html)

**Pull request merged**